shared/sbl.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+type TypeFilter = (domain: string, selector: string, data: Array<*>) => ?boolean
+type TypeDomain = {
+  selectors: {[string]: Function},
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = /^([^/]+)(.+)/.exec(path)
+  var domain = domains[name]
+  if (!domain) {
+    // a call might happen to a selector for domains that haven't been registered
+    // yet, due to module load order. So we save them and run them later.
+    if (!pendingCalls[name]) pendingCalls[name] = []
+    pendingCalls[name].push([path, data])
+    return
+  }
+  // Filters can perform additional functions, and by returning `false` they
+  // can prevent the execution of a selector.
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
+  domains[domain] = {selectors, data: {}, domainFilters: [], selectorFilters: {}}
+  // run any pendingCalls for this domain
+  if (pendingCalls[domain]) {
+    for (var [path, data] of pendingCalls[domain]) {
+      sbl(path, ...data)
+    }
+    delete pendingCalls[domain] // see sbl.throwIfPending
+  }
+}
+
+sbl.addDomainFilter = function (domain: string, filter: TypeFilter) {
+  domains[domain].domainFilters.push(filter)
+}
+
+sbl.setSelectorFilter = function (domain: string, selector: string, filter: TypeFilter) {
+  if (domains[domain].selectorFilters[selector]) {
+    throw new Error(`${domain}/${selector} filter exists`)
+  }
+  domains[domain].selectorFilters[selector] = filter
+}
+
+sbl.addGlobalFilter = function (filter: TypeFilter) {
+  globalFilters.push(filter)
+}
+
+// call at end of sbl setup, right before main run loop of app
+sbl.throwIfPending = function () {
```

Comment:
> calls could be invalid at any point in the application's life-cycle(?)

Yes, that is true. I could fix this by overriding the `.push` method on `pendingCalls` after `throwIfPending` is called (to throw an exception). Is that a change you'd like me to make?

> Further, if we make a call to an api we expect to be synchronous (unless that call expects no return data) then saving a pending call to be run later also suggests that the application has all async calls

Not really... this pending stuff only happens at startup, as modules are loaded, and before anything serious happens. That shouldn't be an issue.

> I think it would be better if you return an error at the time of calling a missing api because that not only identifies whether or not its a typo but also tells where the typo was created without having to encode a complex error text from an object.

That is how it worked originally, but then you get an error because calls are made to domains that haven't been registered yet. Try checking out my branch and making that change to see for yourself.

Without this pendingCalls thing, it becomes necessary to have a dedicated file where all domains are registered, but it would be bad to put all logic for all domains into that one file.

So then you would have to get into obnoxious manual ordering of the module/file load order to prevent errors.